src/charm.py

```diff
@@ -418,6 +446,52 @@ def _on_tls_certificate_pki_certificate_available(self, event: CertificateAvaila
         """Handle the tls-certificates-pki certificate available event."""
         self._add_ca_certificate_to_pki_secrets_engine()
 
+    def _on_vault_pki_certificate_creation_request(
+        self, event: CertificateCreationRequestEvent
+    ) -> None:
+        """Handle the vault-pki certificate creation request event."""
+        self._generate_pki_certificate_for_requirer(
+            event.certificate_signing_request, event.relation_id
+        )
+
+    def _generate_pki_certificate_for_requirer(self, csr: str, relation_id: int):
+        """Generate a PKI certificate for a TLS requirer."""
+        if not self.unit.is_leader():
+            logger.debug("Only leader unit can handle a vault-pki certificate request")
+            return
+        if not self._is_peer_relation_created():
+            logger.debug("Peer relation not created")
+            return
+        vault = self._get_initialized_vault_client()
+        if not vault:
+            logger.debug("Failed to get initialized Vault")
+            return
+        common_name = self._get_config_common_name()
+        if not common_name:
+            logger.error("Common name is not set in the charm config")
+            return
+        if not vault.is_pki_role_created(role=PKI_ROLE, mount=PKI_MOUNT):
+            logger.debug("PKI role not created")
+            return
+        requested_csr = csr
+        requested_common_name = get_common_name_from_csr(requested_csr)
+        if not is_requested_common_name_allowed(requested_common_name, common_name):
+            logger.warning("Requested common name [%s] is not allowed", requested_common_name)
+            return
```

Comment:
I do not think this makes complete sense, as vault would probably have a common name like `vault.example.org`, and I would expect requirers to have names like `req1.example.org`.

Unless there is a hard requirement from Vault itself, I think I would drop this completely and use `tls-constraints` instead, or provide a way to configure policies in Vault.

Reply:
Ok I removed it. It means however that we may regularly make calls to the vault using the client to request for the same csr to be signed. Every time we will handle the exception and log this as a warning.